drivers/mfd: it8801: Implement ITE IT8801 mfd drivers

GTLin08 opened the pull request:
This PR Implements the ITE IT8801 IO expander multi-function device driver. The IO expander provides GPIO, PWM, and keyboard functions via the I2C bus.

GTLin08 committed:
drivers: mfd: it8801: Initialize IT8801 multi-function device drivers

The IT8801 is an I/O expander that provides GPIO, PWM, Keyboard
functions via the I2C bus.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
drivers: input: it8801: Add I2C-based keyboard matrix scan controller

Add I2C-based keyboard matrix scan device driver.
IT8801 support 8 KSI pins and 19 KSO pins [22:11] [6:0].

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
drivers: gpio: it8801: Add I2C-based GPIO device driver

Add I2C-based GPIO device driver. Supports 16-port GPIO divided
into 3 groups.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
drivers: pwm: it8801: Add I2C-based PWM device driver

Add I2C-based PWM device driver. Supports 7 open-drain/push-pull
outputs.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
drivers: mfd: it8801_altctrl: Add alternate controller for MFD

IT8801 support GPIO alternate function switching.
Some GPIO pins can be switched as KSO or PWM function.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

keith-zephyr commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,151 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mfd_ite_it8801, CONFIG_MFD_LOG_LEVEL);
+
+static const struct device *kbd_dev = DEVICE_DT_GET_ANY(ite_it8801_kbd);
```
This driver supports multiple instances, but this this assumes that there is only one IT8801 defined.

GTLin08 replied:
Fixed, use multiple instances.

keith-zephyr commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,151 @@
+	for (i = 0; i < config->sub_mfd_node_num; i++) {
+		if (config->sub_mfd_dev[i] != NULL) {
+			if (config->sub_mfd_dev[i] == kbd_dev) {
+				it8801_input_alert_handler(config->sub_mfd_dev[i]);
```
This function isn't available in this commit.  Each commit should build on it's own.   Add this support with the KBD driver commit.

You could have the child drivers register an alert callback.  Then MFD driver doesn't need to know the type of each child.

If you wanted the alert callbacks to be configured at build time, you could define an iterable section per it8801 instance.

See `INPUT_CALL_DEFINE` for an example of creating an iterable section for a specific device instance.

GTLin08 replied:
Moved to respective driver.

Because the KBD and GPIO devices of IT8801 share the gpio irq. Consider that if IT8801 is connected to the EC, the GPIO interrupt is enabled in the interrupt configuration using irq_connect_dynamic.
Repeated registration of gpio irq will cause [Assert ](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/common/dynamic_isr.c#L23)problem, so here we only register it once in mfd_ite_it8801.

keith-zephyr commented on dts/riscv/ite/it8801-mfd-gpiocr.dtsi:
```diff
@@ -0,0 +1,27 @@
+
+/ {
+	soc {
+		gpio0: gpiocr@0a {
```
Nodename `gpio0` is likely to conflict the the GPIO controller on the SoC.

GTLin08 replied:
Renamed.

keith-zephyr commented on drivers/input/input_ite_it8801_kbd.c:
```diff
@@ -0,0 +1,230 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(input_ite_it8801_kbd, CONFIG_INPUT_LOG_LEVEL);
+
+#define IT8801_MFDCTRL_COUNT IT8801_DT_INST_MFDCTRL_LEN(0)
```
This assumes only one KBD driver installed.  This should be calculated on each KBD instance.

GTLin08 replied:
Fixed, use multiple instances.

keith-zephyr commented on drivers/input/input_ite_it8801_kbd.c:
```diff
@@ -0,0 +1,230 @@
+
+static struct kbd_it8801_data kbd_it8801_data_0;
+
+PM_DEVICE_DT_INST_DEFINE(0, input_kbd_matrix_pm_action);
```
Drivers should support multiple instances.

GTLin08 replied:
Fixed, use multiple instances.

keith-zephyr commented on drivers/pwm/pwm_ite_it8801.c:
```diff
@@ -0,0 +1,183 @@
+	static struct pwm_it8801_data pwm_it8801_data_##inst;                                      \
+	DEVICE_DT_INST_DEFINE(inst, &pwm_it8801_init, NULL, &pwm_it8801_data_##inst,               \
+			      &pwm_it8801_cfg_##inst, POST_KERNEL,                                 \
+			      CONFIG_PWM_IT8801_INIT_PRIORITY, &pwm_it8801_api);
```
Add a `BUILD_ASSERT` that the init priority is lower (higher number)  than the MFD parent.

fabiobaltieri replied:
I suggested in a separate comment that this could all use the same priority, then the sub priority would take care of sequencing the init correctly, I think it's a better approach, less options kicking around. Build asserts should not be needed anymore since the build priorites are checked at build time btw.

GTLin08 replied:
Is there anything I can refer to here? I'm encountering a build error when using CONFIG_MFD_INIT_PRIORITY + 1 in Kconfig or here.

GTLin08 replied:
Use `CONFIG_MFD_INIT_PRIORITY `on all the child devices.

keith-zephyr commented on drivers/gpio/Kconfig.it8801:
```diff
@@ -0,0 +1,22 @@
+	int "IT8801 GPIO port init priority"
+	default 85
+	help
+	  IT8801 GPIO port device driver initialization priority. The priority
+	  must be lower than MFD_INIT_PRIORITY device.
```
Add a BUILD_ASSERT to check the init priority.

GTLin08 replied:
Use CONFIG_MFD_INIT_PRIORITY on all the child devices.

keith-zephyr reviewed (changes requested):
Also consider running clang-format on all the files.

fabiobaltieri commented on dts/riscv/ite/it8801-mfd-gpiocr.dtsi:
```diff
@@ -0,0 +1,27 @@
+ */
+
+/ {
+	soc {
```
This does not look quite right, how is this dtsi going to be used? the child nodes here should be child nodes of the mfd device, right?

GTLin08 replied:
Yes, this is the child node that mfd will use. Renamed.

fabiobaltieri commented on drivers/pwm/pwm_ite_it8801.c:
```diff
@@ -0,0 +1,183 @@
+
+struct pwm_it8801_data {
+	/* I2C device for the MFD parent */
+	const struct i2c_dt_spec *i2c_dev;
```
`ite,it8801-pwm` is a child node of the mfd device right? Then can you get this with `DEVICE_DT_GET(DT_PARENT(...))` at build time?

GTLin08 replied:
That's right, thanks for the tip, we can use this: I2C_DT_SPEC_GET(DT_INST_PARENT(inst))

fabiobaltieri reviewed (commented):
Just an quick initial, pass, noticed some comments of mine overlaps with Keith.

Hey is there a board where you could add this? I'm a bit confused by the dtsi file you have in one of the commits, would be nice to have build testing in CI as well.

GTLin08 commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,151 @@
+	struct mfd_it8801_data *data = dev->data;
+	int ret;
+
+	if (!device_is_ready(config->i2c_dev.bus)) {
```
Done.

GTLin08 commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,151 @@
+		.i2c_dev = I2C_DT_SPEC_INST_GET(inst),                                             \
+		.irq_gpios = GPIO_DT_SPEC_INST_GET(inst, irq_gpios),                               \
+		.sub_mfd_dev = sub_mfd_dev_##inst,                                                 \
+		.sub_mfd_node_num = DT_INST_CHILD_NUM(inst),                                       \
```
Done.

GTLin08 commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,151 @@
+	int i;
+
+	for (i = 0; i < config->sub_mfd_node_num; i++) {
+		if (config->sub_mfd_dev[i] != NULL) {
```
Redundant, removed.

GTLin08 commented on drivers/mfd/Kconfig.it8801:
```diff
@@ -9,3 +9,15 @@ config MFD_ITE_IT8801
+	bool "ITE IT8801 MFD alternate controller"
+	default y
+
+config MFD_IT8801_ALTCTRL_INIT_PRIORITY
+	int "IT8801 GPIO port init priority"
+	default 81
```
Is there anything I can refer to here? I'm encountering a build error when using CONFIG_MFD_INIT_PRIORITY + 1  here.

GTLin08 replied:
Thank you for the information, fixed.

GTLin08 commented:
> Just an quick initial, pass, noticed some comments of mine overlaps with Keith.
> 
> Hey is there a board where you could add this? I'm a bit confused by the dtsi file you have in one of the commits, would be nice to have build testing in CI as well.

I have created a test program to connect with it8xxx2_evb as shown in the [link](https://chromium-review.googlesource.com/c/chromiumos/third_party/zephyr/+/5860875)

fabiobaltieri commented:
> I have created a test program to connect with it8xxx2_evb as shown in the [link](https://chromium-review.googlesource.com/c/chromiumos/third_party/zephyr/+/5860875)

Could you make some equivalent ones in `tests/drivers/build_all/input/app.overlay`, `tests/drivers/build_all/pwm/boards/native_sim.overlay` and `tests/drivers/build_all/gpio/app.overlay`?

fabiobaltieri commented on drivers/gpio/gpio_ite_it8801.c:
```diff
@@ -0,0 +1,486 @@
+		return ret;
+	}
+
+	return ret;
```
`return 0`, can't be anything else, removes any ambiguity

GTLin08 replied:
Done.

fabiobaltieri commented on drivers/gpio/gpio_ite_it8801.c:
```diff
@@ -0,0 +1,486 @@
+
+	*value = val;
+
+	return ret;
```
`return 0`, same in the other below

GTLin08 replied:
Done.

fabiobaltieri commented on drivers/gpio/gpio_ite_it8801.c:
```diff
@@ -0,0 +1,486 @@
+				       bool set)
+{
+	struct gpio_it8801_data *data = dev->data;
+	int rc;
```
`int ret`, for coherency with the rest of the file

GTLin08 replied:
Done.

fabiobaltieri commented on drivers/gpio/gpio_ite_it8801.c:
```diff
@@ -0,0 +1,486 @@
+	int ret;
+	uint8_t reg_gpcr = config->reg_gpcr + pin;
+	uint8_t mask = BIT(pin);
+	uint8_t new_value = 0, control;
```
I think it'd be cleaner if you were to not initialize this one here and instead add a `else` case below that explicitly sets it to 0, then all possible assignment would be near each other

GTLin08 replied:
Done.

fabiobaltieri commented on drivers/gpio/gpio_ite_it8801.c:
```diff
@@ -0,0 +1,486 @@
+	}
+
+	/* Disable irq before configuring it  */
+	ret = i2c_reg_update_byte_dt(&config->i2c_dev, config->reg_gpier, mask, new_value);
```
can you use `0` here for the last argument and avoid initializing `new_value`?

GTLin08 replied:
Done.

fabiobaltieri commented on drivers/gpio/gpio_ite_it8801.c:
```diff
@@ -0,0 +1,486 @@
+	}
+
+	/* Clear pending interrupt */
+	new_value = mask;
```
can you use mask directly? (if you use the `0` above)

GTLin08 replied:
Done.

GTLin08 commented:
Rebased.

GTLin08 commented on drivers/pwm/pwm_ite_it8801.c:
```diff
@@ -0,0 +1,174 @@
+LOG_MODULE_REGISTER(pwm_ite_it8801, CONFIG_PWM_LOG_LEVEL);
+
+struct it8801_pwm_map {
+	int pushpull_en;
```
Done.

GTLin08 committed:
tests: drivers: build_all: Add IT8801 mfd drivers test

Add build tests for the IT8801 MFD drivers, including GPIO,
Input and PWM functionalities.

GPIO, Input test: west build -p always -b native_sim
PWM test: west build -p always -b it82xx2_evb

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 commented:
> > I have created a test program to connect with it8xxx2_evb as shown in the [link](https://chromium-review.googlesource.com/c/chromiumos/third_party/zephyr/+/5860875)
> 
> Could you make some equivalent ones in `tests/drivers/build_all/input/app.overlay`, `tests/drivers/build_all/pwm/boards/native_sim.overlay` and `tests/drivers/build_all/gpio/app.overlay`?

I'm going to create this. Is this part also included in this PR?

fabiobaltieri commented:
> I'm going to create this. Is this part also included in this PR?

That would be ideal yeah, that way the code gets built from the go.

GTLin08 commented:
> > I'm going to create this. Is this part also included in this PR?
> 
> That would be ideal yeah, that way the code gets built from the go.

Done.

GTLin08 commented:
Rebase and resolve conflicting files.

GTLin08 commented:
Hi @fabiobaltieri, @keith-zephyr, If there’s any further feedback on the updates I made. Thanks!

keith-zephyr commented on tests/drivers/build_all/pwm/boards/it82xx2_evb.overlay:
```diff
@@ -0,0 +1,41 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "../../../../../dts/riscv/ite/it8801-mfd-map.dtsi"
```
Is the relative path necessary?  Does `#include <ite/it8801-mfd-map.dtsi>` work instead?

GTLin08 replied:
When using the PWM of IT8801, we must select the alternate pin of the GPIO. The` mfdctrl = <&pwm7_gp20_default>` in dts is defined in "it8801-mfd-map.dtsi", so we need to Include this file.

Fixed to `#include <ite/it8801-mfd-map.dtsi>`

keith-zephyr commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,143 @@
+		 * will scan all the subdevices, and call the subdevice handler
+		 * to determine which device was triggered.
+		 */
+#ifdef CONFIG_INPUT_ITE_IT8801_KBD
+		it8801_input_alert_handler(config->sub_mfd_dev[i]);
+#endif
+#ifdef CONFIG_GPIO_ITE_IT8801
+		it8801_gpio_alert_handler(config->sub_mfd_dev[i]);
+#endif
```
Another option here.  The children devices already have a reference to the parent `mfd`.  The MFD driver could add an API to register an interrupt callback.

```
mfd_it8801_register_interrupt_callback(const struct device *mfd,
    const struct device *child,
    callback_function);
```
Then you could also drop `sub_mfd_dev` array, and you could drop the `if (dev != data->gpio_dev)` checks in the child drivers.

GTLin08 replied:
Thanks for the suggestion! It really helped simplify the code.

GTLin08 commented on dts/riscv/ite/it8801-mfd-gpiocr.dtsi:
```diff
@@ -0,0 +1,30 @@
+ */
+
+/ {
+	it8801_mfd {
```
Done.

keith-zephyr commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,131 @@
+{
+	struct mfd_it8801_data *data = CONTAINER_OF(work, struct mfd_it8801_data, gpio_isr_worker);
+
+	data->mfd_cb(data->sub_mfd_dev);
```
This should check that `mfd_cb` has been assigned.

GTLin08 replied:
Done.

keith-zephyr commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,131 @@
+	k_work_submit(&data->gpio_isr_worker);
+}
+
+void mfd_it8801_register_interrupt_callback(const struct device *mfd, const struct device *child,
+					    it8801_callback_handler_t cb)
```
Is there always a 1:1 relationship from the MFD parent to the child that needs an interrupt callback?
Or does the MFD need to maintain a list of callbacks registered?

GTLin08 replied:
Yes, the MFD needs to maintain a list of registered callbacks, and I have made the modifications.

keith-zephyr commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,154 @@
+{
+	struct mfd_it8801_data *data = mfd->data;
+
+	/* Save the sub node device of mfd for the callback function */
+	if (share_irq_name == share_irq_kbd) {
+		data->sub_dev_kbd = child;
+		data->mfd_cb[share_irq_kbd] = cb;
+	} else {
+		data->sub_dev_gpio[gpio_inst][idx] = child;
+		data->mfd_cb[share_irq_gpio] = cb;
```
This is still more complicated than needed. The MFD parent doesn't need to know if the child is a KBD, GPIO, or PWM.

You should create a structure that each child is responsible for allocating:
```
struct mfd_callback {
    it8801_callback_handler_t cb;
    const struct device *dev;
};
```
Then the `mfd_it8801_register_interrupt_callback()` routine needs to just add the `mfd_callback` allocated by the child to a linked list.  You can use the the `gpio_manage_callback()` for creating the linked list.

Then when dispatching the interrupt in `it8801_gpio_alert_worker`, you only need to iterate over the linked list.  No special checks for KBD vs GPIO are needed.

GTLin08 replied:
Done.

fabiobaltieri commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,132 @@
+		}
+
+		if (val != it8801_id_verify[i].chip_id) {
+			LOG_ERR("The IT8801 vendor ID is wrong");
```
log `i` and `val` with the error?

GTLin08 replied:
Done.

fabiobaltieri commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,132 @@
+		}
+	}
+
+	return ret;
```
`return 0`

GTLin08 replied:
Done.

fabiobaltieri commented on drivers/input/input_ite_it8801_kbd.c:
```diff
@@ -0,0 +1,219 @@
+	if (col == INPUT_KBD_MATRIX_COLUMN_DRIVE_NONE) {
+		/* KSO[22:11, 6:0] output high */
+		kso_val = IT8801_REG_MASK_KSOSDIC | IT8801_REG_MASK_AKSOSC;
+	}
+	/* Assert all outputs */
+	else if (col == INPUT_KBD_MATRIX_COLUMN_DRIVE_ALL) {
```
should be `} else if {`, maybe move the comment after the condition and indent it?

GTLin08 replied:
Done. Also make the comments above and below consistent.

fabiobaltieri commented on drivers/input/input_ite_it8801_kbd.c:
```diff
@@ -0,0 +1,219 @@
+		ret = i2c_reg_write_byte_dt(&config->i2c_dev, config->reg_ksiier, 0x00);
+	}
+
+	if (ret != 0) {
+		LOG_ERR("Failed to set detect mode (ret %d)", ret);
+		return;
+	}
```
this should really be after each call since for the `enable == true` case you have two, just copy paste a couple times :-)

GTLin08 replied:
Done.

fabiobaltieri commented on drivers/input/input_ite_it8801_kbd.c:
```diff
@@ -0,0 +1,219 @@
+	ret = i2c_reg_write_byte_dt(&config->i2c_dev, IT8801_REG_GIECR, IT8801_REG_MASK_GKSIIE);
+	/* Alert response enable */
+	ret = i2c_reg_write_byte_dt(&config->i2c_dev, IT8801_REG_SMBCR, IT8801_REG_MASK_ARE);
+	if (ret != 0) {
+		LOG_ERR("Failed to initialization setting (ret %d)", ret);
+		return ret;
+	}
```
same, add it after every conditions above, keep the message the same so it does not eat more flash, these are unlikely to hit anyway

GTLin08 replied:
Done.

fabiobaltieri reviewed (commented):
cool, few more comments, sorry for the delay this fell of my radar

fabiobaltieri commented on drivers/mfd/mfd_it8801_altctrl.c:
```diff
@@ -0,0 +1,66 @@
+	}
+
+	/* Common settings for alternate function. */
+	ret = i2c_reg_update_byte_dt(i2c_dev, reg_gpiocr, GENMASK(7, 6), alt_val << 6);
```
can you add a define for `GENMASK(7, 6)` and `6`? something with the register name in it I guess

GTLin08 commented:
Rebased.

fabiobaltieri commented on drivers/input/input_ite_it8801_kbd.c:
```diff
@@ -0,0 +1,228 @@
+	return (~value) & 0xff;
+}
+
+void it8801_input_alert_handler(const struct device *dev)
```
`static`

fabiobaltieri commented on drivers/gpio/gpio_ite_it8801.c:
```diff
@@ -0,0 +1,479 @@
+	return ret;
+}
+
+void it8801_gpio_alert_handler(const struct device *dev)
```
`static`

GTLin08 replied:
Done.

fabiobaltieri reviewed (commented):
cool, couple missing static qualifiers, rest is good, great job, Keith is out for the next couple of weeks I think it's fine to wait for him?

GTLin08 commented on drivers/mfd/mfd_ite_it8801.c:
```diff
@@ -0,0 +1,134 @@
+	return 0;
+}
+
+void it8801_gpio_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
```
Done.

GTLin08 commented:
> cool, couple missing static qualifiers, rest is good, great job, Keith is out for the next couple of weeks I think it's fine to wait for him?

OKey, thank you for your feedback!

GTLin08 commented on drivers/gpio/gpio_ite_it8801.c:
```diff
@@ -0,0 +1,479 @@
+	interrupted_pins_level |= (value & data->level_high_trig);
+	interrupted_pins_level |= ((~value) & data->level_low_trig);
+
+	if ((isr_val & ier_val) || interrupted_pins_level) {
```
The IT8801 already supports level-triggered interrupts internally. Therefore, the `level_high_trig `and `level_low_trig `fields are redundant and have been removed.
